### PR TITLE
Add taint upstream authority

### PR DESCRIPTION
**Pull request by MarcosDY**

### CODE REVIEW

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 1]
```diff
@@ -192,6 +236,7 @@ message RevokeX509AuthorityResponse {
         AuthorityState revoked_authority = 1;
 }
 
+
```
Comment (amartinezfayo) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:239]: Unnecessary blank line?

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 2]
```diff
@@ -62,13 +62,30 @@ service LocalAuthority {
         // will perform proactive rotations of any key material related to
         // the tainted authority. The result of this action will be observed
         // cluster-wide.
-        // It can receive the public key of an old X.509 authority.
+        // It can receive the authority ID of an old X.509 authority.
+	// 
+	// If upstream authority is configured, local authorities can not be tainted,
+	// a FailedPrecondition error will be returned.
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:68]:
```suggestion
	// If an upstream authority is configured then local authorities cannot be tainted,
	// and a FailedPrecondition error will be returned.
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 3]
```diff
@@ -62,13 +62,30 @@ service LocalAuthority {
...
         //
         // If a previously active X.509 authority does not exist (e.g. if one
         // has been prepared but not activated yet), a FailedPrecondition
         // error will be returned.
         rpc TaintX509Authority(TaintX509AuthorityRequest) returns (TaintX509AuthorityResponse);
 
+        // TaintX509UpstreamAuthority marks the provided upstream authority as
+        // being tainted. SPIRE Agents observing an authority to be tainted
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:76]:
```suggestion
        // being tainted. SPIRE Agents observing a tainted authority 
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 4]
```diff
@@ -62,13 +62,30 @@ service LocalAuthority {
...
+        // will perform proactive rotations of any key material related to
+        // the tainted authority. The result of this action will be observed
+        // cluster-wide.
+	// It is important to change active upstream authority before taiting it,
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:80]:
```suggestion
	// It is important to change to a new active upstream authority before tainting the old one, 
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 5]
```diff
@@ -62,13 +62,30 @@ service LocalAuthority {
...
 
+        // TaintX509UpstreamAuthority marks the provided upstream authority as
+        // being tainted. SPIRE Agents observing an authority to be tainted
+        // will perform proactive rotations of any key material related to
+        // the tainted authority. The result of this action will be observed
+        // cluster-wide.
+	// It is important to change active upstream authority before taiting it, 
+	// since tainting will force the rotation of any bundle that is using 
+	// the old upstream authority.
+        // It receives the X.509 Subject Key Identifier (or SKID) of an old X.509 authority.
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:83]:
```suggestion
        // The X.509 authority to taint is identified using the provided X.509 Subject Key Identifier (or SKID) of the old X.509 authority.
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 6]
```diff
@@ -62,13 +62,30 @@ service LocalAuthority {
...
+        //
+        // If an X.509 upstream authority does not exist or it is active, a FailedPrecondition
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:85]:
```suggestion
        // If an X.509 upstream authority is not configured, or the identified upstream X.509 authority is active, a FailedPrecondition
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 7]
```diff
@@ -78,6 +95,16 @@ service LocalAuthority {
         // has been prepared but not activated yet), a FailedPrecondition
         // error will be returned.
         rpc RevokeX509Authority(RevokeX509AuthorityRequest) returns (RevokeX509AuthorityResponse);
+
+        // RevokeX509UpstreamAuthority revokes the previously active X.509 upstream authority by
+        // removing it from the bundle and propagating this update throughout
+        // the cluster.
+        // It receives the subject key ID of the authority's CA certificate of the
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:102]:
```suggestion
        // The X.509 authority to revoke is identified using the provided subject key ID of the authority's CA certificate.
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 8]
```diff
@@ -62,13 +62,31 @@ service LocalAuthority {
         // will perform proactive rotations of any key material related to
         // the tainted authority. The result of this action will be observed
         // cluster-wide.
-        // It can receive the public key of an old X.509 authority.
+        // It can receive the authority ID of an old X.509 authority.
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:65]:
```suggestion
        // The X.509 authority to taint is identified using the provided X.509 Subject Key
	// Identifier (or SKID) of the old X.509 authority.
```

File: proto/spire/api/server/localauthority/v1/localauthority.proto [hunk 9]
```diff
@@ -62,13 +62,31 @@ service LocalAuthority {
...
+	// If an upstream authority is configured then local authorities cannot be tainted,
+	// and a FailedPrecondition error will be returned.
         //
         // If a previously active X.509 authority does not exist (e.g. if one
         // has been prepared but not activated yet), a FailedPrecondition
         // error will be returned.
         rpc TaintX509Authority(TaintX509AuthorityRequest) returns (TaintX509AuthorityResponse);
 
+        // TaintX509UpstreamAuthority marks the provided upstream authority as
+        // being tainted. SPIRE Agents observing a tainted authority to be tainted
```
Comment (azdagron) [inline proto/spire/api/server/localauthority/v1/localauthority.proto:76]:
```suggestion
        // being tainted. SPIRE Agents observing a tainted authority

**Review by azdagron — APPROVED** [review]

**Review by amartinezfayo — APPROVED** [review]

**Pull request merged**